### ISSUE TITLE
KeyError on server side , then  i can't connect the server

### ISSUE THREAD

**Issue by mat7213** [hanxiao/bert-as-service#162]
Process BertSink-1:
Traceback (most recent call last):
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib/python3.6/site-packages/bert_serving/server/__init__.py", line 226, in run
    job_checksum[job_id]))
KeyError: b'd4853427-8e43-48ee-8b8f-a24eb357face#1'

After this ,   i can't connect the server to get sentence encode , how to solve it?
release:
     python3.6
     pip list：
     bert-serving-client 1.5.0     
     bert-serving-server 1.5.0 
server  is started by : 
    nohup bert-serving-start -model_dir ../multi_cased_L-12_H-768_A-12 -num_worker=8  -cpu &

**Comment by hanxiao** [comment]
version too old, please first do

pip install -U bert-serving-server bert-serving-client

if the problem still arises, then please report it here again.

D.D.Conan <notifications@github.com>于2018年12月26日 周三下午2:49写道：

**Comment by mat7213** [comment]
after i do 
pip install -U bert-serving-server bert-serving-client
version now is 
  bert-serving-client 1.6.4     
  bert-serving-server 1.6.4

then i start server :
-bash-4.2$  bert-serving-start -model_dir ../multi_cased_L-12_H-768_A-12 -num_worker=8  -cpu
Traceback (most recent call last):
  File "/bin/bert-serving-start", line 9, in <module>
    from bert_serving.server.bert.extract_features import PoolingStrategy
ImportError: cannot import name 'PoolingStrategy'

is there missing any python package to install ?

**Comment by hanxiao** [comment]
no, i’m guessing you manually create ‘/bin/bert-serving-start’, which still
links to the old version. Please check.

D.D.Conan <notifications@github.com>于2018年12月26日 周三下午3:09写道：

**Comment by mat7213** [comment]
yes, update bert-serving-start solved the ImportError
but about an hour later , KeyError again...

I:VENTILATOR:[__i:_ru:149]:new encode request   req id: 1       size: 1 client: b'0d516f0e-a988-4257-bf6e-29fb33589ba2'
Process BertSink-2:
Traceback (most recent call last):
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/home/appuser/.local/lib/python3.6/site-packages/bert_serving/server/__init__.py", line 225, in run
    self._run()
  File "/usr/lib64/python3.6/site-packages/zmq/decorators.py", line 75, in wrapper
    return func(*args, **kwargs)
  File "/usr/lib64/python3.6/site-packages/zmq/decorators.py", line 75, in wrapper
    return func(*args, **kwargs)
  File "/usr/lib64/python3.6/site-packages/zmq/decorators.py", line 75, in wrapper
    return func(*args, **kwargs)
  File "/home/appuser/.local/lib/python3.6/site-packages/bert_serving/server/__init__.py", line 264, in _run
    job_checksum[job_id]))
KeyError: b'01f59169-0beb-4afb-a3dd-6932c43c6169#1'

**Comment by hanxiao** [comment]
I need more information about how you use `BertClient`.  Looks like some corner case that can not be reproduced accurately.

1. Is there any reason why you encode **one sentence a time**? Looks like you do! I said clearly in the README that encoding one sentence a time is a VERY VERY BAD idea: https://github.com/hanxiao/bert-as-service#speed-wrt-client_batch_size
I hope you are not creating a new `BertClient()` after every sentence, that's even worse. 

2. When you encounter the `KeyError`, e.g. `KeyError: b'01f59169-0beb-4afb-a3dd-6932c43c6169#1'`, could you search the server log for `01f59169-0beb-4afb-a3dd-6932c43c6169`? What did the `VENTILATOR`, `WORKER`, `SINK` say about this id? Could you copy paste the relevant lines here?

3. Did you use `BertClient.encode()` or `BertClient.encode_async()`?

**Comment by mat7213** [comment]
1、sorry...but i need to  encode one sentence a time , and create a new `BertClient()` after every request . Because i'm developing a service  to compare one new sentence with base sentences , which likes a  simple chatbot.
2、the full log:
`I:VENTILATOR[__i:run:133]:new config request        req id: 0       client: b'01f59169-0beb-4afb-a3dd-6932c43c6169'
I:VENTILATOR[__i:run:133]:new config request        req id: 0       client: b'0d516f0e-a988-4257-bf6e-29fb33589ba2'
I:SINK[__i:run:286]:send config     client b'01f59169-0beb-4afb-a3dd-6932c43c6169'
I:VENTILATOR[__i:run:149]:new encode request        req id: 1       size: 1 client: b'01f59169-0beb-4afb-a3dd-6932c43c6169'
I:WORKER-6[__i:gen:384]:new job     socket: 0       size: 1 client: b'01f59169-0beb-4afb-a3dd-6932c43c6169#1'
I:WORKER-6[__i:run:364]:job done    size: (1, 768)  client: b'01f59169-0beb-4afb-a3dd-6932c43c6169#1'
I:SINK[__i:run:286]:send config     client b'0d516f0e-a988-4257-bf6e-29fb33589ba2'
I:VENTILATOR[__i:run:149]:new encode request        req id: 1       size: 1 client: b'0d516f0e-a988-4257-bf6e-29fb33589ba2'
Process BertSink-2:
Traceback (most recent call last):
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/home/appuser/.local/lib/python3.6/site-packages/bert_serving/server/__init__.py", line 225, in run
    self._run()
  File "/usr/lib64/python3.6/site-packages/zmq/decorators.py", line 75, in wrapper
    return func(*args, **kwargs)
  File "/usr/lib64/python3.6/site-packages/zmq/decorators.py", line 75, in wrapper
    return func(*args, **kwargs)
  File "/usr/lib64/python3.6/site-packages/zmq/decorators.py", line 75, in wrapper
    return func(*args, **kwargs)
  File "/home/appuser/.local/lib/python3.6/site-packages/bert_serving/server/__init__.py", line 264, in _run
    job_checksum[job_id]))
KeyError: b'01f59169-0beb-4afb-a3dd-6932c43c6169#1'`

3、While every request to my service , i  would create  bc = BertClient() , use bc.encode() to get sentence encode , then bc.close() : 
`    def get_recommend(self,question):
        bc = BertClient()
        test_sentence = question
        sentence_list = []
        sentence_list.append(test_sentence)
        test_vec =  bc.encode(sentence_list)
        bc.close()
        ……
`

**Comment by hanxiao** [comment]
why can't you persist one BertClient and reuse it for every `get_recommand`?

```python
class YourClass:
  def __init__(self):
   self.bc = BertClient()

  def get_recommend(self, question):
    test_vec = bc.encode([question])
```

Why can't you do this?

**Comment by mat7213** [comment]
Thanks !  I have modified the code.
It's a Django  service ,  i was afraid  of causing  long-term connection  problems by persisting one BertClient in init part.

**Comment by hanxiao** [comment]
Good to know. 

Some technical details here: if you are worry about "long-term connection" in `BertClient`, then please  don't. The `BertClient` does not persist a connection with server. It pushes the data and subscribes the result independently. This is designed on purpose to make the overall system more tangible and reliable, e.g. for server restart.

One example you could try is starting the BertServer, then making a BertClient; then killing and restarting the server. Now using **the same** BertClient to communicate with server. Guess what, still works!

Why? If you look at the communication flow here https://github.com/hanxiao/bert-as-service#q-what-is-the-parallel-processing-model-behind-the-scene

![image](https://user-images.githubusercontent.com/2041322/50445934-f3905e00-094c-11e9-8e47-48b123d9e9e3.png)

Note, the client is using PUSH and SUBSCRIBE for communicating with the server. You can imagine them as "one-time" protocol, meaning no delay no wait for response, and are setup only on-demand.

**Comment by hanxiao** [comment]
despite your imperfect way of using BertClient before, you pointed out a bug on the server side. It is a very corner case, but yes it is a bug. It is solved in #164
